File: r/R/oura.R
# Oura Ring API v2 provider
#
# Wraps https://api.ouraring.com/v2 — uses TokenBridge for auth,
# handles cursor-based pagination internally.
#
# Usage:
#
#   tb_set_provider("oura")
#   tb_fetch("p001", "sleep", start, end)          # canonical
#
#   ou_fetch("p001", "daily-activity", start, end)  # Oura shorthand
#
#   tok <- tb_get_token("p001", provider = "oura")
#   ou_fetch("p001", "sleep",           start, end, token = tok)
#   ou_fetch("p001", "daily-readiness", start, end, token = tok)
#
# Data type IDs: see names(OU_DATA_TYPES) or docs/providers/index.md
#
# Function prefix: ou_   (Oura)

# ── Data types reference ───────────────────────────────────────────────────────

#' All supported Oura data type IDs
#'
#' Named list describing the API path and date parameter format for each type.
#' Use \code{names(OU_DATA_TYPES)} to list all available type IDs.
#'
#' Full descriptions and field details: see
#' \url{https://cloud.ouraring.com/v2/docs}
#'
#' @export
OU_DATA_TYPES <- list(
  # Daily summaries
  "daily-activity"     = list(path = "daily_activity",           date_fmt = "date"),
  "daily-sleep"        = list(path = "daily_sleep",              date_fmt = "date"),
  "daily-readiness"    = list(path = "daily_readiness",          date_fmt = "date"),
  "daily-stress"       = list(path = "daily_stress",             date_fmt = "date"),
  "daily-spo2"         = list(path = "daily_spo2",               date_fmt = "date"),
  "daily-resilience"   = list(path = "daily_resilience",         date_fmt = "date"),
  "cardiovascular-age" = list(path = "daily_cardiovascular_age", date_fmt = "date"),
  # Detailed sessions
  "sleep"              = list(path = "sleep",                    date_fmt = "date"),
  "sleep-time"         = list(path = "sleep_time",               date_fmt = "date"),
  "workout"            = list(path = "workout",                  date_fmt = "date"),
  "session"            = list(path = "session",                  date_fmt = "date"),
  # Continuous streams
  "heartrate"          = list(path = "heartrate",                date_fmt = "datetime"),
  # Other
  "vo2-max"            = list(path = "vO2_max",                  date_fmt = "date"),
  "tag"                = list(path = "tag",                      date_fmt = "date"),
  "enhanced-tag"       = list(path = "enhanced_tag",             date_fmt = "date")
)

# ── Public API ─────────────────────────────────────────────────────────────────

#' Fetch Oura Ring data for a participant
#'
#' @param user_id    TokenBridge user ID
#' @param data_type  Data type ID (see \code{names(OU_DATA_TYPES)})
#' @param start_date "YYYY-MM-DD"
#' @param end_date   "YYYY-MM-DD"
#' @param token      Pre-fetched access token (skips TokenBridge round-trip)
#' @param env_file   Path to .env file (default ".env")
#' @return data.frame, one row per record
#' @export
ou_fetch <- function(user_id, data_type, start_date, end_date,
                     token = NULL, env_file = ".env") {
  .ou_validate_dates(start_date, end_date)

  if (!data_type %in% names(OU_DATA_TYPES)) {
    stop(
      "Unknown Oura data type: ", shQuote(data_type),
      ". Available: ", paste(names(OU_DATA_TYPES), collapse = ", "),
      call. = FALSE
    )
  }

  if (is.null(token)) {
    token <- tb_get_token(user_id, provider = "oura", env_file = env_file)
  }

  .ou_fetch_all(token, data_type, start_date, end_date)
}

# ── Internal ───────────────────────────────────────────────────────────────────

.ou_validate_dates <- function(start_date, end_date) {
  fmt <- "%Y-%m-%d"
  s <- tryCatch(as.Date(start_date, format = fmt), warning = function(w) NA)
  e <- tryCatch(as.Date(end_date,   format = fmt), warning = function(w) NA)

  if (is.na(s) || is.na(e) ||
      format(s, fmt) != start_date || format(e, fmt) != end_date) {
    stop("Dates must be YYYY-MM-DD format, got: ",
         shQuote(start_date), ", ", shQuote(end_date), call. = FALSE)
  }
  if (s > e) {
    stop("start_date (", start_date, ") must not be after end_date (", end_date, ")",
         call. = FALSE)
  }
}

.ou_to_iso_datetime <- function(date_str, end_of_day = FALSE) {
  d <- as.POSIXct(date_str, tz = "UTC")
  if (end_of_day) d <- d + 86399L
  format(d, "%Y-%m-%dT%H:%M:%S+00:00")
}

.ou_fetch_all <- function(token, data_type, start_date, end_date) {
  spec    <- OU_DATA_TYPES[[data_type]]
  url     <- paste0("https://api.ouraring.com/v2/usercollection/", spec$path)
  headers <- httr::add_headers(Authorization = paste("Bearer", token))

  if (spec$date_fmt == "datetime") {
    date_params <- list(
      start_datetime = .ou_to_iso_datetime(start_date),
      end_datetime   = .ou_to_iso_datetime(end_date, end_of_day = TRUE)
    )
  } else {
    date_params <- list(start_date = start_date, end_date = end_date)
  }

  all_records <- list()
  next_token  <- NULL

  repeat {
    params <- date_params
    if (!is.null(next_token)) params$next_token <- next_token

    resp <- httr::GET(url, headers, query = params)

    if (httr::status_code(resp) != 200L) {
      stop("Oura API HTTP error: ", httr::status_code(resp), call. = FALSE)
    }

    body        <- httr::content(resp, as = "parsed", type = "application/json")
    records     <- body$data %||% list()
    all_records <- c(all_records, records)
    next_token  <- body$next_token

    if (is.null(next_token)) break
  }

  if (length(all_records) == 0L) return(data.frame())

  as.data.frame(
    do.call(rbind, lapply(all_records, function(r) {
      as.data.frame(lapply(r, function(v) if (length(v) == 1) v else list(v)),
                    stringsAsFactors = FALSE)
    })),
    stringsAsFactors = FALSE
  )
}

`%||%` <- function(a, b) if (!is.null(a)) a else b

File: python/tokenbridge/providers/oura.py
"""Oura Ring API v2 provider.

Wraps https://api.ouraring.com/v2 — uses TokenBridge for auth,
handles pagination internally.

All responses are flat lists of dicts matching the Oura API field names
exactly (no transformation needed — Oura returns clean JSON already).

Supported data types::

    from tokenbridge.providers.oura import DATA_TYPES
    print(list(DATA_TYPES))

See https://cloud.ouraring.com/v2/docs for field-level details.
"""

from datetime import date, datetime, timezone
from typing import Optional, Sequence

import requests

from tokenbridge.providers._base import HealthProvider

_BASE = "https://api.ouraring.com/v2/usercollection"

# date_fmt: "date" → start_date/end_date (YYYY-MM-DD)
#           "datetime" → start_datetime/end_datetime (ISO 8601 with timezone)
DATA_TYPES: dict[str, dict] = {
    # ── Daily summaries ───────────────────────────────────────────────────────
    "daily-activity":       {"path": "daily_activity",              "date_fmt": "date"},
    "daily-sleep":          {"path": "daily_sleep",                 "date_fmt": "date"},
    "daily-readiness":      {"path": "daily_readiness",             "date_fmt": "date"},
    "daily-stress":         {"path": "daily_stress",                "date_fmt": "date"},
    "daily-spo2":           {"path": "daily_spo2",                  "date_fmt": "date"},
    "daily-resilience":     {"path": "daily_resilience",            "date_fmt": "date"},
    "cardiovascular-age":   {"path": "daily_cardiovascular_age",    "date_fmt": "date"},
    # ── Detailed sessions ─────────────────────────────────────────────────────
    "sleep":                {"path": "sleep",                       "date_fmt": "date"},
    "sleep-time":           {"path": "sleep_time",                  "date_fmt": "date"},
    "workout":              {"path": "workout",                     "date_fmt": "date"},
    "session":              {"path": "session",                     "date_fmt": "date"},
    # ── Continuous streams ────────────────────────────────────────────────────
    "heartrate":            {"path": "heartrate",                   "date_fmt": "datetime"},
    # ── Other ─────────────────────────────────────────────────────────────────
    "vo2-max":              {"path": "vO2_max",                     "date_fmt": "date"},
    "tag":                  {"path": "tag",                         "date_fmt": "date"},
    "enhanced-tag":         {"path": "enhanced_tag",                "date_fmt": "date"},
}


def _validate_dates(start_date: str, end_date: str) -> None:
    fmt = "%Y-%m-%d"
    try:
        s = datetime.strptime(start_date, fmt).date()
        e = datetime.strptime(end_date, fmt).date()
    except ValueError:
        raise ValueError(
            f"Dates must be YYYY-MM-DD format, got: {start_date!r}, {end_date!r}"
        )
    if s > e:
        raise ValueError(
            f"start_date ({start_date}) must not be after end_date ({end_date})"
        )


def _to_iso_datetime(date_str: str, end_of_day: bool = False) -> str:
    """Convert YYYY-MM-DD to ISO 8601 datetime with UTC timezone."""
    d = date.fromisoformat(date_str)
    if end_of_day:
        dt = datetime(d.year, d.month, d.day, 23, 59, 59, tzinfo=timezone.utc)
    else:
        dt = datetime(d.year, d.month, d.day, tzinfo=timezone.utc)
    return dt.isoformat()


class Oura(HealthProvider):
    """Oura Ring API v2 provider.

    Usually accessed via ``tb.oura`` rather than instantiated directly.

    All public methods accept an optional ``token`` argument:

    .. code-block:: python

        token = tb.get_token("p001", provider="oura")
        sleep = ou.fetch("p001", "sleep",          start, end, token=token)
        ready = ou.fetch("p001", "daily-readiness", start, end, token=token)
    """

    PROVIDER_ID = "oura"

    def fetch(
        self,
        user_id: str,
        data_type: str,
        start_date: str,
        end_date: str,
        *,
        token: Optional[str] = None,
    ) -> list[dict]:
        """Fetch any Oura data type by its ID.

        Args:
            user_id: TokenBridge participant ID.
            data_type: Type ID from :data:`DATA_TYPES`, e.g. ``"sleep"``,
                ``"daily-activity"``, ``"heartrate"``.
            start_date: Start of date range, ``"YYYY-MM-DD"``.
            end_date: End of date range, ``"YYYY-MM-DD"``.
            token: Pre-fetched access token.  Obtain with
                ``tb.get_token(user_id, provider="oura")``.

        Returns:
            List of dicts, one per record.  Field names match the Oura API
            exactly.  Returns an empty list if no data exists for the period.

        Raises:
            ValueError: If ``data_type`` is not in :data:`DATA_TYPES` or
                dates are invalid.
        """
        _validate_dates(start_date, end_date)
        if data_type not in DATA_TYPES:
            raise ValueError(
                f"Unknown Oura data type {data_type!r}. "
                f"Available: {sorted(DATA_TYPES)}"
            )
        if token is None:
            token = self._get_token(user_id)
        return _fetch_all(token, data_type, start_date, end_date)


# ── Internal fetch helpers ────────────────────────────────────────────────────

def _fetch_all(token: str, data_type: str, start_date: str, end_date: str) -> list[dict]:
    spec     = DATA_TYPES[data_type]
    url      = f"{_BASE}/{spec['path']}"
    headers  = {"Authorization": f"Bearer {token}"}

    if spec["date_fmt"] == "datetime":
        params: dict = {
            "start_datetime": _to_iso_datetime(start_date),
            "end_datetime":   _to_iso_datetime(end_date, end_of_day=True),
        }
    else:
        params = {"start_date": start_date, "end_date": end_date}

    records: list[dict] = []
    next_token: Optional[str] = None

    while True:
        if next_token:
            params["next_token"] = next_token

        resp = requests.get(url, params=params, headers=headers, timeout=30)
        resp.raise_for_status()
        body = resp.json()

        records.extend(body.get("data", []))

        next_token = body.get("next_token")
        if not next_token:
            break

    return records
